### ISSUE TITLE
Auto infer type parameters not given from function's arguments

### ISSUE THREAD

**Issue by xyzingh** [microsoft/TypeScript#39780]
## Search Terms

infer, derivation, deduction, type, param, parameter, function

### Background

I'm a library developer and need to provide a function with 1 required parameter and 2 type parameters.

```ts
type UtilType<T, U> = ...;
function foo<T, U>(required: U): UtilType<T, U> {
    // ...
}
```

Users can choose to provide the type `T` to refine the function's return type `UtilType<T,U>`, or not.

Meanwhile, type `U` should never need to be specified manually, but automatically inferred from parameter `required`, which is usually an long object literal.

But it doesn't work as well as I thought.

```ts
foo({ abc: 123 }); // not specify T. OK, work well.
foo<string>({ abc: 123 }); // specify T. Error: Expected 2 type parameters, but got 1.
```

So I change the source code, let `U` have a default value.

```ts
type UtilType<T, U> = ...;
function foo<T, U = any>(required: U): UtilType<T, U> {
    // ...
}

// It still doesn't work well.

foo<string>({ abc: 123 });
// The way TS understand:  function foo<string, any>(required: any): UtilType<string, any>
```

### Solution

One solution that won't bring breaking change is:

If TS find a required type param that is not given but can be inferred from the parameters of the function, use that type and do not report compilation errors any more.

```ts
type UtilType<T, U> = ...;
function foo<T, U>(required: U): UtilType<T, U> {
    // ...
}
foo<string>({ abc: 123 }); // T = string, U = { abc: number }
```

Another solution:

Currently, when the user does not give any type parameter, the type inferred from function's arguments has a higher priority than the default type.

```ts
function bar<T = string, U = number>(arg: U) { }
bar(true); // TS think:  function bar<string, boolean>(arg: boolean): void
```

But when there is any type specified, TS's behavior changed strangely.

```ts
function bar<T = string, U = number>(arg: U) { }
bar<string>(true); // Error: true cannot be assigned to number
```

This solution is to always acknowledge that the inferred type has a higher priority than the default type, and this may cause breaking change. But I think this is more graceful.

## Examples

Solution 1:

```ts
type UtilType<T, U> = ...;
function foo<T, U>(required: U): UtilType<T, U> {
    // ...
}
foo<string>({ abc: 123 }); // T = string, U = { abc: number }
// No compilation errors any more
```

Solution 2:

```ts
type UtilType<T, U> = ...;
function foo<T, U = any>(required: U): UtilType<T, U> {
    // ...
}
foo<string>({ abc: 123 }); // T = string, ** U = { abc: number } **
// inferred type has a higher priority than default type
```

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by MartinJohns** [comment]
It sounds like you want #26242. Currently only either all type arguments can be inferred, or none. #26242 would allow to infer only specific type arguments.

**Comment by xyzingh** [comment]
@MartinJohns  Thank you. I think the discussion of #26242 is more comprehensive and in-depth than this. I will close this issue.
